src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -0,0 +1,51 @@
+require 'puppet_pal'
+require 'puppet/server/log_collector'
+require 'puppet/server/logging'
+
+module Puppet
+  module Server
+    class ASTCompiler
+      def self.compile(compile_options)
+        options = compile_options['options'] || {}
+
+        log_level = options['log_level']
+
+        if options['capture_logs']
+          catalog, logs = Logging.capture_logs(log_level) do
+            compile_ast(compile_options)
+          end
+
+          { catalog: catalog, logs: logs }
+        else
+          catalog = compile_ast(compile_options)
+          { catalog: catalog }
+        end
+      end
+
+      def self.compile_ast(compile_options)
+        # FromDataConverter.convert below expects this to be a hash
+        code = JSON.parse(compile_options['code_ast'])
+        target = compile_options['certname']
+
+        # Use the existing environment with the requested name
+        Puppet::Pal.in_environment(compile_options['environment'],
+                                   envpath: Puppet[:environmentpath],
+                                   facts: compile_options['facts']['values'],
+                                   variables: compile_options['variables']['values']) do |pal|
+          Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
+
+          # This compiler has been configured with a node containing
+          # the requested environment, facts, and variables, and is used
+          # to compile a catalog in that context from the supplied AST.
+          pal.with_catalog_compiler do |compiler|
+            ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
+            compiler.evaluate(ast)
+            compiler.compile_additions
+            compiler.with_json_encoding(&:encode)
```

Comment:
this feels like a weird interface, but I guess it has nothing to do with this pr